### PR TITLE
Improve behavior of Disabled plugins

### LINKED ISSUE #4778: v2.18 regression in behavior of `APOLLO_SCHEMA_REPORTING=true` + `engine: {reportSchema: false}`
Before the big refactor in v2.18, setting `engine: {reportSchema: false}` would override the env var `APOLLO_SCHEMA_REPORTING=true`.

In v2.18 the `engine` parameter is just for backwards compatibility. By accident, this combination now leads to schema reporting; there is no way in code to override `APOLLO_SCHEMA_REPORTING=true`.

(Note: both of these statements are based on code observation, not running code.)

Two things that perhaps can be done here:
- Introduce a `ApolloServerPluginSchemaReportingDisabled`, like the other `*Disabled` plugins, that overrides the env var.
- Change the legacy `engine` option handling so that `false` overrides the env var.

Workaround: Don't set `APOLLO_SCHEMA_REPORTING` in environments where you don't want schema reporting; either only set it in appropriate contexts or only enable schema reporting via `plugins: [ApolloServerPluginSchemaReporting()]`.

### LINKED ISSUE #7099: Consider warning/error if "Disabled" plugin combined with explicit installation
We received reports of users confused that the effect of (essentially):

```
new ApolloServer({
  plugins: [
    ApolloServerPluginUsageReporting({ some: 'custom configuration' }),
    ApolloServerPluginUsageReportingDisabled(),
  ],
})
```

is to leave usage reporting on despite the "disabled" plugin. We should consider just making this into an error (which would be sorta-backwards-incompatible but only sorta, this was not intended to have well-defined semantics, and AS4 is new anyway), or perhaps a warning although warnings easily just become spam.

(The "disabled" plugins actually mean "disable the automatic installation of the corresponding plugin with default configuration" but currently has no effect if you manually install the corresponding plugin.)

### PR DESCRIPTION
**Pull request by glasser**
- Make it an error to install a built-in plugin and its "Disabled" counterpart in the same server. "Disabled" really means "don't install by default", but it's reasonable to assume it would counteract a manual installation of the same plugin too.  For simplicity, this makes installing both flavors into an error at start time. This is technically backwards-incompatible but we are still early in the ASv4 lifecycle and this combination wasn't expected to have well-defined behavior.
- Add `ApolloServerPluginSchemaReportingDisabled` which allows you to override the effect of the `APOLLO_SCHEMA_REPORTING` environment variable.
- Doc improvements.

Fixes #4778. Fixes #7099.

### CODE REVIEW

File: docs/source/api/apollo-server.mdx [hunk 1]
```diff
@@ -372,10 +372,10 @@ To learn more about configuring Apollo Server's cache, see [Configuring cache ba
...
-
 You can _also_ add plugins to your server before you start it using the [`addPlugin`](#addplugin) method.
 
+Apollo Server comes with several plugins that it installs automatically in their default configuration if certain conditions are met. For example, the usage reporting plugin is installed if you provide a graph API key and a graph ref. Apollo Server skips this automatic installation if manually provide the plugin (in the `plugins` array or with the `addPlugin` method), or if you provide the plugin's corresponding "disabled" plugin (such as `ApolloServerPluginUsageReportingDisabled()` for `ApolloServerPluginUsageReporting`). For details, see the API references for these plugins: [usage reporting](./plugin/usage-reporting/), [schema reporting](./plugin/schema-reporting/), [landing page](./plugin/landing-pages/), [cache control](./plugin/cache-control/), and [inline trace](./plugin/inline-trace/).
```
Comment (trevor-scheer) [inline docs/source/api/apollo-server.mdx:377]:
```suggestion
Apollo Server comes with several plugins that it installs automatically in their default configuration if certain conditions are met. For example, the usage reporting plugin is installed if you provide a graph API key and a graph ref. Apollo Server skips this automatic installation if you manually provide the plugin (in the `plugins` array or with the `addPlugin` method), or if you provide the plugin's corresponding "disabled" plugin (such as `ApolloServerPluginUsageReportingDisabled()` for `ApolloServerPluginUsageReporting`). For details, see the API references for these plugins: [usage reporting](./plugin/usage-reporting/), [schema reporting](./plugin/schema-reporting/), [landing page](./plugin/landing-pages/), [cache control](./plugin/cache-control/), and [inline trace](./plugin/inline-trace/).
```

File: packages/server/src/ApolloServer.ts [hunk 2]
```diff
@@ -807,6 +807,49 @@ export class ApolloServer<in out TContext extends BaseContext = BaseContext> {
...
+    ] of pluginsByInternalID.entries()) {
+      if (sawDisabled && sawNonDisabled) {
+        throw new Error(
+          `You have tried to install both ApolloServerPlugin${id} and ` +
+            `ApolloServerPlugin${id}Disabled in your server. Please choose ` +
+            `whether or not you want to disable the feature and install the ` +
+            `appropriate plugin for your use case.`,
+        );
+      }
+    }
```
Comment (trevor-scheer) [inline packages/server/src/ApolloServer.ts:851]: Pro: this is verbose and it's easy to understand the logic here.

Con: we don't really need this loop, and could even cut the first one short and throw as soon as we see the true/true condition.

File: packages/server/src/internalPlugin.ts [hunk 3]
```diff
@@ -12,6 +12,7 @@ export interface InternalApolloServerPlugin<TContext extends BaseContext>
   // Used to identify a few specific plugins that are instantiated
   // by default if not explicitly used or disabled.
   __internal_plugin_id__(): InternalPluginId;
+  __is_disabled_plugin__(): boolean;
```
Comment (trevor-scheer) [inline packages/server/src/internalPlugin.ts:15]: nbd, and this matches the existing pattern, but is it helpful that this is a function and not just a boolean?

**Review by trevor-scheer — APPROVED** [review]

File: packages/server/src/internalPlugin.ts (hunk 3 above)
Comment (glasser) [reply packages/server/src/internalPlugin.ts:15]: I just assumed I had a reason for doing it this way before with the other one :)  Will change both

File: packages/server/src/ApolloServer.ts (hunk 2 above)
Comment (glasser) [reply packages/server/src/ApolloServer.ts:851]: Sure.

**Pull request merged**